modules/photo/src/merge.cpp

```diff
@@ -171,95 +173,117 @@ class MergeMertensImpl CV_FINAL : public MergeMertens
+                img.convertTo(img, CV_32F, 1.0f/255.0f);
+                if(channels == 3) {
+                    cvtColor(img, gray, COLOR_RGB2GRAY);
+                } else {
+                    img.copyTo(gray);
+                }
+                split(img, splitted);
 
-            Mat mean = Mat::zeros(size, CV_32F);
-            for(int c = 0; c < channels; c++) {
-                mean += splitted[c];
-            }
-            mean /= channels;
+                Laplacian(gray, contrast, CV_32F);
+                contrast = abs(contrast);
 
-            saturation = Mat::zeros(size, CV_32F);
-            for(int c = 0; c < channels;  c++) {
-                Mat deviation = splitted[c] - mean;
-                pow(deviation, 2.0f, deviation);
-                saturation += deviation;
-            }
-            sqrt(saturation, saturation);
+                Mat mean = Mat::zeros(size, CV_32F);
+                for(int c = 0; c < channels; c++) {
+                    mean += splitted[c];
+                }
+                mean /= channels;
 
-            wellexp = Mat::ones(size, CV_32F);
-            for(int c = 0; c < channels; c++) {
-                Mat expo = splitted[c] - 0.5f;
-                pow(expo, 2.0f, expo);
-                expo = -expo / 0.08f;
-                exp(expo, expo);
-                wellexp = wellexp.mul(expo);
-            }
+                saturation = Mat::zeros(size, CV_32F);
+                for(int c = 0; c < channels;  c++) {
+                    Mat deviation = splitted[c] - mean;
+                    pow(deviation, 2.0f, deviation);
+                    saturation += deviation;
+                }
+                sqrt(saturation, saturation);
 
-            pow(contrast, wcon, contrast);
-            pow(saturation, wsat, saturation);
-            pow(wellexp, wexp, wellexp);
+                wellexp = Mat::ones(size, CV_32F);
+                for(int c = 0; c < channels; c++) {
+                    Mat expo = splitted[c] - 0.5f;
+                    pow(expo, 2.0f, expo);
+                    expo = -expo / 0.08f;
+                    exp(expo, expo);
+                    wellexp = wellexp.mul(expo);
+                }
+
+                pow(contrast, wcon, contrast);
+                pow(saturation, wsat, saturation);
+                pow(wellexp, wexp, wellexp);
+
+                weights[i] = contrast;
+                if(channels == 3) {
+                    weights[i] = weights[i].mul(saturation);
+                }
+                weights[i] = weights[i].mul(wellexp) + 1e-12f;
 
-            weights[i] = contrast;
-            if(channels == 3) {
-                weights[i] = weights[i].mul(saturation);
+                std::unique_lock<std::mutex> lock(weight_sync.mutex);
+                weight_sync.cond.wait(lock, [&]{ return weight_sync.index == i; });
+                weight_sync.sum += weights[i];
+                weight_sync.index++;
+                weight_sync.cond.notify_all();
```

Comment:
> `weight_sync.index == i`

Do we really need to preserve summation operations order?

If no, then please use `cv::AutoLock` and `cv::Mutex` instead (to keep working `OPENCV_DISABLE_THREAD_SUPPORT` mode)

Also manual usage of cond variables is not exception-safe.

Reply:
> Do we really need to preserve summation operations order?

I wasn't sure what the policy for OpenCV is, so my first attempt was to preserve deterministic results.

> If no, then please use `cv::AutoLock` and `cv::Mutex` instead (to keep working `OPENCV_DISABLE_THREAD_SUPPORT` mode)

If fine with the policy, then we can of course also remove the condition variable and use OpenCV's Mutex/AutoLock. I changed it accordingly, PTAL!
I adjusted the python test_umat_merge_mertens though to only use one thread for testing, as we want to have determinism for the summation here.